sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/JobFailure.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.loadtests;
+
+import static java.lang.String.format;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+
+import java.io.IOException;
+import java.util.Optional;
+import org.apache.beam.sdk.PipelineResult;
+import org.joda.time.Duration;
+
+/** Class for detecting failures after {@link PipelineResult#waitUntilFinish(Duration)} unblocks. */
+class JobFailure {
+
+  private String cause;
+
+  private boolean requiresCancelling;
+
+  JobFailure(String cause, boolean requiresCancelling) {
+    this.cause = cause;
+    this.requiresCancelling = requiresCancelling;
+  }
+
+  static Optional<JobFailure> assertFailure(PipelineResult pipelineResult,
+      LoadTestResult testResult) {
+    PipelineResult.State state = pipelineResult.getState();
+
+    Optional<JobFailure> stateRelatedFailure = lookForInvalidState(state);
+
+    if (stateRelatedFailure.isPresent()) {
+      return stateRelatedFailure;
+    } else {
+      return lookForMetricResultFailure(testResult);
+    }
+  }
+
+  private static Optional<JobFailure> lookForMetricResultFailure(LoadTestResult testResult) {
+    if (testResult.getRuntime() == -1 || testResult.getTotalBytesCount() == -1) {
+      return of(new JobFailure("Invalid test results", false));
+    } else {
+      return empty();
+    }
+  }
+
+  static Optional<JobFailure> lookForInvalidState(PipelineResult.State state) {
+    switch (state) {
+      case RUNNING:
+      case UNKNOWN:
+        return of(new JobFailure("Job timeout.", true));
+
+      case CANCELLED:
+      case FAILED:
+      case STOPPED:
+      case UPDATED:
+        return of(new JobFailure(format("Invalid job state: %s.", state.toString()), false));
+
+      default:
+        return empty();
+    }
+  }
+
+  static PipelineResult handleFailure(PipelineResult pipelineResult, JobFailure failure)
+      throws IOException {
+
+    if(failure.requiresCancelling) {
+      pipelineResult.cancel();
+    }
+    throw new RuntimeException(failure.cause);
```

Comment:
Is there a way to get a full stacktrace from the runner rather than the status (failed) only? Ideally, I'd like to display full error in Jenkins logs without the need to dig into google cloud console/other runner's console.